Onboard Diagnostics

wlan0 opened the pull request:
on board diagnostics

@krishnasrinivas @poornas @abperiasamy 

[![asciicast](https://asciinema.org/a/309756.svg)](https://asciinema.org/a/309756)

The output from the ascii video is here: https://gist.github.com/wlan0/670db2881b4163e3af35a94f83bf642a

Some interesting observations to not in the json above:
 - The speed diff between minio-4, which has nvme, and others, which only have hdds. 
 - The 100 Gbit network link is saturated
 - Differences between the CPUs of different machines

Here's another output JSON of a fs setup with 1 drive on my laptop. It's smaller, so its easier to read

https://gist.github.com/wlan0/aa892c51c126dfed6a47305eddeec967

wlan0 committed:
add obd backend support

wlan0 committed:
disk diagnostic and config format updates

disk diagnostics: preserve diagnostics in memory after 1st run
config format: return map of values instead of bytestream
admin obd command: fix logic for setting query params to ensure all obd types are either true or false

wlan0 committed:
NetOBD info and Admin Info

1. add admin info to obd api
2. add network OBD info
3. lock obd commands to allow only one at a time

poornas commented on cmd/admin-handlers.go:
```diff
@@ -1360,6 +1361,84 @@ func (a adminAPIHandlers) ServerHardwareInfoHandler(w http.ResponseWriter, r *ht
+		}
+		defer obdLock.Unlock()
+
+	}
```
@wlan0, the handler shouldn't have logic specific to fs or xl. Suggest defining a GetOBDInfo function in object-api-interface.go and move the functionality here to xl/fs layers. For gateway, you will just need to  raise a  NotImplemented error in gateway-unsupported.go

wlan0 replied:
I'll do that

wlan0 replied:
done

poornas commented on cmd/admin-router.go:
```diff
@@ -162,6 +162,9 @@ func registerAdminRouter(router *mux.Router, enableConfigOps, enableIAMOps bool)
 	adminRouter.Methods(http.MethodGet).Path(adminAPIVersionPrefix + "/kms/key/status").HandlerFunc(httpTraceAll(adminAPI.KMSKeyStatusHandler))
 
+	// -- OBD API --
+	adminRouter.Methods(http.MethodGet).Path(adminAPIVersionPrefix+"/obdinfo").HandlerFunc(httpTraceAll(adminAPI.OBDInfoHandler)).Queries("drive", "{drive:true|false}", "net", "{net:true|false}", "info", "{info:true|false}", "config", "{config:true|false}")
```
might want to just do a httpTraceHdrs here

wlan0 replied:
done

poornas commented on cmd/admin-server-info.go:
```diff
@@ -90,6 +90,49 @@ func getLocalCPULoad(endpointZones EndpointZones, r *http.Request) ServerCPULoad
+	var drivesOBDInfo []madmin.DriveOBDInfo
+	for _, ep := range endpointZones {
+		for _, endpoint := range ep.Endpoints {
+			// Only proceed for local endpoints
```
```suggestion
                      if !endpoint.IsLocal {
                         continue
                      }
```
and remove the `if endpoint.IsLocal {` to avoid nesting

poornas commented on cmd/notification.go:
```diff
@@ -950,6 +950,78 @@ func (sys *NotificationSys) CollectNetPerfInfo(size int64) map[string][]ServerNe
+		netOBDs[index].Addr = addr
+		if err != nil {
+			netOBDs[index].Error = err.Error()
+		}
```
```suggestion
		netOBDs[index].Addr = addr
		if err != nil {
			netOBDs[index].Error = err.Error()
		        reqInfo := (&logger.ReqInfo{}).AppendTags("remotePeer", addr)
		        ctx := logger.SetReqInfo(context.Background(), reqInfo)
		        logger.LogIf(ctx, err)
		}
```

wlan0 replied:
:+1:

poornas commented on pkg/disk/obd.go:
```diff
@@ -0,0 +1,165 @@
+package disk
```
file is missing license header

poornas reviewed (changes requested):
admin docs will also need an update

wlan0 committed:
OBD Improvements

- Implement graph algorithm to test network bandwidth from every node to every other node
- Saturate any network bandwidth adaptively, accounting for slow and fast network capacity
- Implement parallel drive OBD tests
- Implement a paging mechanism for OBD test to provide periodic updates to client
- Implement mechanism to dynamically cancel long running OBD tests

wlan0 commented on cmd/utils.go:
```diff
@@ -64,6 +64,15 @@ func IsErr(err error, errs ...error) bool {
 	return false
 }
 
+func printStackTraceHere(w io.Writer) {
```
This should be removed

wlan0 committed:
Implement Sys, Process, Host, Mem  OBD Infos

- Sys: Disk (Partitions, DiskInfo) , CPU (HW, Timers)
- Host: Platform (OS, Kernel), Sensors (Temperatures)
- Mem: Swap Mem, Virtual Mem
- Process: Everything about the MinIO Process

klauspost commented on cmd/api-errors.go:
```diff
@@ -799,6 +800,12 @@ var errorCodes = errorCodeMap{
+	ErrOBDLockTimeout: {
+		Code:           "UnsupportedOperation",
+		Description:    "parallel OBD requests are not supported at this time",
+		HTTPStatusCode: http.StatusBadRequest,
```
```suggestion
		HTTPStatusCode: http.http.StatusConflict,
```
More accurate?

wlan0 replied:
:+1:

klauspost commented on cmd/fs-v1.go:
```diff
@@ -226,6 +226,21 @@ func (fs *FSObjects) StorageInfo(ctx context.Context, _ bool) StorageInfo {
 }
 
+// GetOBDLock - returns a locked cluster-wide lock to prevent parallel OBD operations
+func (fs *FSObjects) GetOBDLock(ctx context.Context) (RWLocker, error) {
```
This this is added to the interface, we might as well make it generic.
```suggestion
func (fs *FSObjects) GetRWLock(ctx context.Context, name string, timeout) (RWLocker, error) {
```

wlan0 replied:
:+1:

klauspost commented on cmd/object-api-interface.go:
```diff
@@ -139,4 +139,7 @@ type ObjectLayer interface {
 	DeleteObjectTag(context.Context, string, string) error
+
+	// OBD
+	GetOBDLock(context.Context) (RWLocker, error)
```
If made generic, move up to other lock function at top.

wlan0 replied:
:+1:

klauspost commented on cmd/obdinfo.go:
```diff
@@ -0,0 +1,426 @@
+
+		// errors out if current process doens't have any children
+		children, _ := proc.ChildrenWithContext(ctx)
+		// if err != nil {
```
Uncomment or remove.

klauspost commented on cmd/peer-rest-client.go:
```diff
@@ -190,6 +199,322 @@ func (client *peerRESTClient) NetworkInfo() (info madmin.ServerNetworkHardwareIn
+	buflimiter := make(chan bool, threadCount)
+	errChan := make(chan error, threadCount)
+
+	totalTransferred := int64(0)
```
All of this is kinda iffy.

You keep a totalTransferred across all `maxSamples` goroutines, with threadCount running at once.

I see an intention of trying to determine the average speed while the operation was going, but the `totalTransferred` is an ahead-of-time counter that does not say much about when the transfer is taking place.

The actual readahead of the transfer buffer will apply a significant skew to the numbers AFAICT. 
Furthermore this also relies on the `transferChan` being emptied at once - this could at least be remedied by having the counter inside the `progressReader` and update it at once when the read occurs.

Furthermore, once there is less than `threadCount` transfers running you are basically testing something new and these should be discarded.

wlan0 replied:
Neither is true.

> I see an intention of trying to determine the average speed while the operation was going, but the totalTransferred is an ahead-of-time counter 

`totalTransferred` is not an ahead-of-time counter. `totalTransferred` is only incremented after the buffer has been emptied of `n` bytes.

>   that does not say much about when the transfer is taking place. 

Recording the start and stop time alone is actually the wrong approach. I'll explain why - If you measured start and stop time of `n` operations, you'd end up with n posibbly-overlapping intervals.

```
1-3
2-4
2-6
3-5
...
```

You'll then have to break them apart into non-overlapping intervals and compute the number of active threads in each of those.

```
Interval      |   threads
1-2           |    1
2-3           |    2
3-4           |    3
4-5           |    2
5-6           |    1 
``` 

Using this data, the throughput can only be computed with the assumption that data is being transferred at a constant rate of `partial_size/sec` for a given thread in a given interval. 
where partial_size is computed as 

```
    partial_size = obj_size * number of seconds to complete entire op / number of seconds in interval
```

The throughput of the entire interval can only be computed as the sum of throughputs of each thread in that interval. The issue with this approach is that it assumes the throughput for a given object in a given thread is uniformly distributed. This actually leads to skewness. 

In contrast, my approach measures throughput in terms of total bytes transferred between two points in time. There is simply no need to count the number of active threads or the non-overlapping intervals from each of the individual overlapping operations. This approach, In addition to being simple and elegant, it always leads to accurate results. 

> The actual readahead of the transfer buffer will apply a significant skew to the numbers AFAICT.

There is no readahead. 

> Furthermore this also relies on the transferChan being emptied at once - this could at least be remedied by having the counter inside the progressReader and update it at once when the read occurs.

I don't think you fully understand the approach. Please refer to my explanation above. 

> Furthermore, once there is less than threadCount transfers running you are basically testing something new and these should be discarded.

If I relied on the `threadCount` being emptied at once, the results would be skewed every single time. A simple sanity check will prove otherwise. In my throughput computation, I'm not multiplying the size of an object with the number of active threads. I believe your assumption is coming from the approach in warp and it is bound to skewness. Please refer to my explanation above.

klauspost replied:
> `totalTransferred` is not an ahead-of-time counter. `totalTransferred` is only incremented after the buffer has been emptied of `n` bytes.

Correct, didn't read it correctly.

However, get rid of the lag imposed by the `transferChan` channel. Right now your `before`/`after` relies on everything in `transferChan` to have been applied. But you can just get rid of it.

> In contrast, my approach measures throughput in terms of total bytes transferred between two points in time.

> There is simply no need to count the number of active thread.

My main objection is that this assumes that 1 connection can saturate the NIC. We know that isn't true which is why you have several goroutines running. So we should consider the number of requests, or at least reject the ones at the end where less are running.

wlan0 replied:
Again, neither is true. Maybe we should get on a call, so I can explain this to you better.

klauspost commented on cmd/peer-rest-client.go:
```diff
@@ -190,6 +199,322 @@ func (client *peerRESTClient) NetworkInfo() (info madmin.ServerNetworkHardwareIn
+
+	buf := make([]byte, dataSize)
+
+	buflimiter := make(chan bool, threadCount)
```
```suggestion
	buflimiter := make(chan struct{}, threadCount)
```

wlan0 replied:
:+1:

klauspost commented on cmd/peer-rest-client.go:
```diff
@@ -190,6 +199,322 @@ func (client *peerRESTClient) NetworkInfo() (info madmin.ServerNetworkHardwareIn
+		}
+	}
+
+	wg := &sync.WaitGroup{}
```
No need for a pointer.
```suggestion
	var wg sync.WaitGroup
```

klauspost commented on cmd/peer-rest-server.go:
```diff
@@ -509,6 +510,151 @@ func (s *peerRESTServer) ServerInfoHandler(w http.ResponseWriter, r *http.Reques
+		return
+	}
+
+	ctx := newContext(r, w, "DispatchNetOBDInfo")
```
You can use [keepHTTPResponseAlive](https://github.com/minio/minio/blob/8d9866263334a644ed8a80a427ed7bad792c1dce/cmd/storage-rest-server.go#L611) here and [waitForHTTPResponse](https://github.com/minio/minio/blob/8d9866263334a644ed8a80a427ed7bad792c1dce/cmd/storage-rest-server.go#L639) client side to make sure the connection remains alive.

Maybe also relevant for other calls.

klauspost commented on docs/debugging/README.md:
````diff
@@ -23,3 +23,37 @@ To trace entire HTTP request and also internode communication
+OBD data saved to dc-11-obd_20200321053323.json.gz
+```
+
+The gunzipped output contains debugging information for your system
````
```suggestion
The gzipped output contains debugging information for your system.
```

wlan0 replied:
:+1:

klauspost commented on pkg/disk/obd.go:
```diff
@@ -0,0 +1,182 @@
+const kb = uint64(1 << 10)
+const mb = uint64(kb << 10)
+const gb = uint64(mb << 10)
+
```
```suggestion
const (
	kb = 1 << 10
	mb = kb << 10
	gb = mb << 10
)
```

wlan0 replied:
:+1:

klauspost commented on pkg/disk/obd.go:
```diff
@@ -0,0 +1,182 @@
+	Avg          float64 `json:"avg_secs,omitempty"`
+	Percentile50 float64 `json:"percentile50_secs,omitempty"`
+	Percentile90 float64 `json:"percentile90_secs,omitempty"`
+	Percentile99 float64 `json:"percentile99_secs,omitempty"`
```
Do you have enough information for a 99 percentile?

klauspost commented on pkg/disk/obd.go:
```diff
@@ -0,0 +1,182 @@
+
+// Throughput holds throughput information for write operations to the drive
+type Throughput struct {
+	Avg          float64 `json:"avg_bytes/s,omitempty"`
```
```suggestion
	Avg          float64 `json:"avg_bps,omitempty"`
```

Pretty sure slashes in a JSON name is a bad idea.

wlan0 replied:
it was bps before. The confusion arose between bits/sec and bytes/sec

klauspost replied:
@wlan0 `avg_bytes_per_sec` or something like it then. No `/` pls.

klauspost commented on pkg/disk/obd.go:
```diff
@@ -0,0 +1,182 @@
+	Avg          float64 `json:"avg_bytes/s,omitempty"`
+	Percentile50 float64 `json:"percentile50_bytes/s,omitempty"`
+	Percentile90 float64 `json:"percentile90_bytes/s,omitempty"`
+	Percentile99 float64 `json:"percentile99_bytes/s,omitempty"`
```
Again, likely to be just noise or just equal to `Min`.

wlan0 replied:
No, this captures tail latency really well. This is quite useful tbh

klauspost commented on pkg/disk/obd.go:
```diff
@@ -0,0 +1,182 @@
+
+// Latency holds latency information for write operations to the drive
+type Latency struct {
+	Avg          float64 `json:"avg_secs,omitempty"`
```
For warp I used milliseconds, since it is usually a bit easier to read. Maybe do that?

klauspost commented on pkg/disk/obd.go:
```diff
@@ -0,0 +1,182 @@
+
+// GetOBDInfo about the drive
+func GetOBDInfo(ctx context.Context, endpoint string) (Latency, Throughput, error) {
+	runtime.LockOSThread()
```
Why?

This doesn't avoid preemption - AFAICT it makes it even more likely. But if there is a reason please document it for future reference.

klauspost commented on pkg/net/obd.go:
```diff
@@ -0,0 +1,116 @@
+
+// Throughput holds throughput information for read/write operations to the drive
+type Throughput struct {
+	Avg          float64 `json:"avg_bytes/s,omitempty"`
```
Same as disk.

poornas commented on cmd/gateway/azure/gateway-azure.go:
```diff
@@ -488,6 +488,11 @@ func (a *azureObjects) StorageInfo(ctx context.Context, _ bool) (si minio.Storag
+// GetOBDLock - Not relevant to Azure backend.
+func (a *azureObjects) GetOBDLock(ctx context.Context) (minio.RWLocker, error) {
+	return nil, minio.NotImplemented{}
+}
```
GetOBDLock() can be defined just in gateway-unsupported.go and removed from other gateway specific files. All the gateway objects inherit from minio.GatewayUnsupported
```
func (a GatewayUnsupported)GetOBDLock(ctx context.Context) (minio.RWLocker, error) {
	return nil, minio.NotImplemented{}
}
```

poornas commented on cmd/gateway/b2/gateway-b2.go:
```diff
@@ -227,6 +227,11 @@ func (l *b2Objects) StorageInfo(ctx context.Context, _ bool) (si minio.StorageIn
+func (l *b2Objects) GetOBDLock(ctx context.Context) (minio.RWLocker, error) {
+	return nil, minio.NotImplemented{}
+}
+
```
this function can be removed

poornas commented on cmd/gateway/hdfs/gateway-hdfs.go:
```diff
@@ -217,6 +217,11 @@ func (n *hdfsObjects) StorageInfo(ctx context.Context, _ bool) minio.StorageInfo
+// GetOBDLock - Not relevant to HDFS backend.
+func (n *hdfsObjects) GetOBDLock(ctx context.Context) (minio.RWLocker, error) {
+	return nil, minio.NotImplemented{}
+}
```
this can be removed

poornas commented on cmd/gateway/nas/gateway-nas.go:
```diff
@@ -116,6 +116,11 @@ func (n *nasObjects) StorageInfo(ctx context.Context, _ bool) minio.StorageInfo
+// GetOBDLock - Not relevant to nas backend.
+func (n *nasObjects) GetOBDLock(ctx context.Context) (minio.RWLocker, error) {
+	return nil, minio.NotImplemented{}
+}
```
this can be removed.

poornas commented on cmd/gateway/oss/gateway-oss.go:
```diff
@@ -328,6 +328,11 @@ func (l *ossObjects) StorageInfo(ctx context.Context, _ bool) (si minio.StorageI
+// GetOBDLock - Not relevant to oss backend.
+func (l *ossObjects) GetOBDLock(ctx context.Context) (minio.RWLocker, error) {
+	return nil, minio.NotImplemented{}
+}
```
this function can be removed

poornas commented on cmd/gateway/s3/gateway-s3.go:
```diff
@@ -280,6 +280,11 @@ func (l *s3Objects) StorageInfo(ctx context.Context, _ bool) (si minio.StorageIn
+// GetOBDLock - Not relevant to S3 backend.
+func (l *s3Objects) GetOBDLock(ctx context.Context) (minio.RWLocker, error) {
+	return nil, minio.NotImplemented{}
+}
```
this function can be removed.

poornas commented on cmd/gateway/gcs/gateway-gcs.go:
```diff
@@ -420,6 +420,11 @@ func (l *gcsGateway) StorageInfo(ctx context.Context, _ bool) (si minio.StorageI
+// GetOBDLock - Not relevant to gcs backend.
+func (l *gcsGateway) GetOBDLock(ctx context.Context) (minio.RWLocker, error) {
+	return nil, minio.NotImplemented{}
+}
```
this function can be removed

klauspost reviewed (approved):
Awesome 👍

poornas reviewed (approved):
LGTM

harshavardhana commented on cmd/admin-handlers.go:
```diff
@@ -1372,6 +1373,135 @@ func (a adminAPIHandlers) ServerHardwareInfoHandler(w http.ResponseWriter, r *ht
+	}
+
+	// returns a locked lock
+	lock, err := objectAPI.GetRWLock(deadlinedCtx, "obd", 5*time.Second, 2*time.Second)
```
You don't need to implement GetRWLock we already have NewNSLock() @wlan0

harshavardhana commented on cmd/fs-v1.go:
```diff
@@ -226,6 +226,21 @@ func (fs *FSObjects) StorageInfo(ctx context.Context, _ bool) StorageInfo {
 }
 
+// GetRWLock - returns a locked cluster-wide lock
+func (fs *FSObjects) GetRWLock(ctx context.Context, name string, timeout, minimum time.Duration) (RWLocker, error) {
```
We can use NewNSLock instead of GetRWLock() its an overkill anyways @wlan0

wlan0 replied:
Yeah, i'll fix that everywhere

harshavardhana commented on cmd/notification.go:
```diff
@@ -890,6 +891,290 @@ func (sys *NotificationSys) CollectNetPerfInfo(size int64) map[string][]ServerNe
+		        'x's should be tested, and 'o's should be skipped
+	*/
+	stripPath := func(hostPath string) string {
+		return strings.Split(hostPath, "/")[0]
```
```suggestion
		return strings.Split(hostPath, slashSeparator)[0]
```

harshavardhana commented on cmd/notification.go:
```diff
@@ -890,6 +891,290 @@ func (sys *NotificationSys) CollectNetPerfInfo(size int64) map[string][]ServerNe
+		return strings.Split(hostPath, "/")[0]
+	}
+
+	hostMap := map[string]bool{}
```
use hostSet := set.NewStringSet() and use that with `hostSet.Contains(host)`

harshavardhana commented on cmd/notification.go:
```diff
@@ -890,6 +891,290 @@ func (sys *NotificationSys) CollectNetPerfInfo(size int64) map[string][]ServerNe
+	hostMap := map[string]bool{}
+	for _, ez := range globalEndpoints {
+		for _, e := range ez.Endpoints {
+			host := stripPath(e.Host)
```
if you are only interested in host you can use e.Hostname() - you can avoid stripPath() function altogether

harshavardhana commented on cmd/admin-handlers.go:
```diff
@@ -1372,6 +1373,135 @@ func (a adminAPIHandlers) ServerHardwareInfoHandler(w http.ResponseWriter, r *ht
+	vars := mux.Vars(r)
+
+	if cpu, ok := vars["syscpu"]; ok && cpu == "true" {
+		cpuInfo := getLocalCPUOBDInfo(deadlinedCtx)
```
getLocalCPUOBDInfo() can take the `r.Host` to make sure that when it's in FS or XL mode not in distributed you don't default to `minio` as the hostname.

harshavardhana replied:
Look at getLocalMemUsage()

harshavardhana replied:
This needs to be repeated across all functions

wlan0 replied:
That;s what I started with. However, you end up getting the path as well, when using `r.Host`

harshavardhana replied:
why would you get the path? @wlan0 not able to understand?

wlan0 replied:
I thought `r.Host` includes both hostname and path

harshavardhana commented on cmd/api-errors.go:
```diff
@@ -800,6 +801,12 @@ var errorCodes = errorCodeMap{
 		HTTPStatusCode: http.StatusBadRequest,
 	},
+	ErrOBDLockTimeout: {
+		Code:           "UnsupportedOperation",
```
Custom errors in admin API should be prefixed with `XMinioAdmin` to avoid conflicting with S3 error codes.

wlan0 replied:
Gotcha

harshavardhana commented on cmd/api-errors.go:
```diff
@@ -800,6 +801,12 @@ var errorCodes = errorCodeMap{
+	ErrOBDLockTimeout: {
+		Code:           "UnsupportedOperation",
+		Description:    "parallel OBD requests are not supported at this time",
+		HTTPStatusCode: http.StatusConflict,
```
We can return `http.StatusRequestTimeout` since the error is Timeout? if not rename it as conflict?

harshavardhana replied:
Then we should say `ErrOBDLockTimeout--> ErrAdminOBDConflict`

harshavardhana commented on cmd/object-api-interface.go:
```diff
@@ -139,4 +140,7 @@ type ObjectLayer interface {
 	DeleteObjectTag(context.Context, string, string) error
+
+	// OBD
+	GetRWLock(context.Context, string, time.Duration, time.Duration) (RWLocker, error)
```
Just for OBD lets not add new object layer locking functions NewNSLock is already exposed at Object layer which can be used.

harshavardhana commented on cmd/peer-rest-common.go:
```diff
@@ -30,6 +30,14 @@ const (
 	peerRESTMethodCPULoadInfo                  = "/cpuloadinfo"
```
update the version to indicate that we are bringing in new methods

harshavardhana commented on cmd/rest/client.go:
```diff
@@ -73,6 +73,11 @@ func (c *Client) CallWithContext(ctx context.Context, method string, values url.
 		return nil, &NetworkError{err}
 	}
 
+	final := resp.Header.Get("FinalStatus")
```
Since trailer is set - are you not looking at resp.Trailer?

harshavardhana commented on cmd/server-main.go:
```diff
@@ -75,16 +75,20 @@ EXAMPLES:
      {{.Prompt}} {{.HelpName}} /home/shared
 
-  2. Start distributed minio server on an 32 node setup with 32 drives each, run following command on all the nodes
+  2. Start minio server on a single server with 16 drives
```
is this added back? we on purpose removed lots of examples - we don't need to mention or encourage single node erasure-coded setups. did @abperiasamy ask you add it back?

wlan0 replied:
yeah @abperiasamy wanted me to add this.

harshavardhana commented on cmd/xl-zones.go:
```diff
@@ -216,6 +216,21 @@ func (z *xlZones) StorageInfo(ctx context.Context, local bool) StorageInfo {
 }
 
+// GetRWLock - returns a locked cluster-wide lock
+func (z xlZones) GetRWLock(ctx context.Context, name string, timeout, minimum time.Duration) (RWLocker, error) {
```
all this repeated code can be avoided and just use NewNSLock()

harshavardhana reviewed (changes requested):
Testing right now using `mc` everything looks good - just some idiomatic changes needed.
